### PR TITLE
Be more cautious about handling invalid battery measurements.

### LINKED ISSUE #215: Can't parse axp288_fuel_gauge uevent
I have an Intel Atom table with a AXP288 PMIC that reports the battery information in the following way:
```
/sys/class/power_supply/axp288_fuel_gauge/uevent:

POWER_SUPPLY_NAME=axp288_fuel_gauge
POWER_SUPPLY_STATUS=Discharging
POWER_SUPPLY_PRESENT=1
POWER_SUPPLY_HEALTH=Good
POWER_SUPPLY_VOLTAGE_MAX_DESIGN=4200000
POWER_SUPPLY_VOLTAGE_NOW=3872000
POWER_SUPPLY_VOLTAGE_OCV=3942000
POWER_SUPPLY_CURRENT_NOW=-928000
POWER_SUPPLY_CAPACITY=75
POWER_SUPPLY_CAPACITY_ALERT_MIN=0
POWER_SUPPLY_TECHNOLOGY=Li-ion
POWER_SUPPLY_CHARGE_FULL=6404944
POWER_SUPPLY_CHARGE_NOW=4816448
```

This causes that batt_info (at [this](https://github.com/i3/i3status/blob/94651257cea4ff419f9cf2143d93193d5a5ccb96/src/print_battery_info.c#L179) point) to be:
```
{ full_design = -3,
  full_last = 25133000,
  remaining = 20505191,
  present_rate = 3327552,
  seconds_remaining = -1,
  percentage_remaining = -1,
  status = CS_DISCHARGING }
```
Which then causes the code to go through [this](https://github.com/i3/i3status/blob/94651257cea4ff419f9cf2143d93193d5a5ccb96/src/print_battery_info.c#L502) path, that causes a "No battery" result.

Can't this be done in a better way?

### PR DESCRIPTION
**Pull request by kousu**
This fixes #215 for me. The (Linux kernel 4.12) axp288 uevent doesn't have `POWER_SUPPLY_CHARGE_FULL_DESIGN` nor `POWER_SUPPLY_ENERGY_FULL_DESIGN`:

```
[kousu@phi ~]$ cat /sys/class/power_supply/axp288_fuel_gauge/uevent 
POWER_SUPPLY_NAME=axp288_fuel_gauge
POWER_SUPPLY_STATUS=Discharging
POWER_SUPPLY_PRESENT=1
POWER_SUPPLY_HEALTH=Good
POWER_SUPPLY_VOLTAGE_MAX_DESIGN=4200000
POWER_SUPPLY_VOLTAGE_NOW=3766000
POWER_SUPPLY_VOLTAGE_OCV=3854000
POWER_SUPPLY_CURRENT_NOW=-976000
POWER_SUPPLY_CAPACITY=67
POWER_SUPPLY_CAPACITY_ALERT_MIN=0
POWER_SUPPLY_TECHNOLOGY=Li-ion
POWER_SUPPLY_CHARGE_FULL=z
POWER_SUPPLY_CHARGE_NOW=6093360
```

When this hit
```
    int full = (last_full_capacity ? batt_info.full_last : batt_info.full_design);
    if (full < 0 && batt_info.percentage_remaining < 0) {
```
it would interpret the missing `.full_design` as a missing battery.

This instead, at length --- and if you can shorten that section I'll applaud you, but it's 4am and I can't see a way that is shorter while still being obviously correct in all cases --- falls back on `.full_capacity` if `.full_design` is missing.

This also allows the existence of `.remaining` to be as good as `.percentage_remaining`, which seems like it's probably related to #216, and adds protection in the ampere to watt conversion section.

### CODE REVIEW

**Comment by danielotero** [comment]
The problem with our AXP288 kernel driver is that it doesn't report the `POWER_SUPPLY_{CHARGE|ENERGY}_FULL_DESIGN=` that i3status expects by default. This behavior can be changed with the `last_full_capacity=true` option, which uses the `POWER_SUPPLY_{CHARGE|ENERGY}_FULL` attribute instead.

The question is: should i3status show the `last_full_capacity=true` version if there is no `POWER_SUPPLY_{CHARGE|ENERGY}_FULL_DESIGN=` attribute? If so, I think the patch could be as simple as:

```Patch
@@ -499,6 +499,9 @@ void print_battery_info(yajl_gen json_gen, char *buffer, int number, const char
             return;
     }
 
+    // If there is no POWER_SUPPLY_*_FULL attribute, fallback to design capacity
+    if (batt_info.full_last < 0)
+        last_full_capacity = true;
     int full = (last_full_capacity ? batt_info.full_last : batt_info.full_design);
     if (full < 0 && batt_info.percentage_remaining < 0) {
         /* We have no physical measurements and no estimates. Nothing
```

**Comment by kousu** [comment]
I did end up shortening it just now, to

```
-    int full = (last_full_capacity ? batt_info.full_last : batt_info.full_design);
+    int full = (batt_info.full_design < 0 || (last_full_capacity && batt_info.full_last >= 0)) ?  batt_info.full_last : batt_info.full_design;
```

which does almost the same thing as you patch, but also handles the case where we have `.full_design` but not `.full_last`.

File: testcases/014-battery-capacity/expected_output.txt [hunk 1]
```diff
@@ -1 +1 @@
-Touchpad: 100.00% BAT 
\ No newline at end of file
+Touchpad: 100.00% BAT
```
Comment (eplanet) [inline testcases/014-battery-capacity/expected_output.txt]: I took the liberty of including that fix in PR #239.

**Comment by kousu** [comment]
Hi there, I think I've dealt with all the code reviews I've gotten. Any idea on when/if this will get merged ^^?

**Comment by Airblader** [comment]
@stapelberg Can you take over? I'll actually be on vacation as of today, so I won't be doing much over the next week. :-)

**Comment by stapelberg** [comment]
> @stapelberg Can you take over? I'll actually be on vacation as of today, so I won't be doing much over the next week. :-)

Of course. Enjoy your vacation!

**Comment by stapelberg** [comment]
Could you squash your commits into logical entities before we merge this please? I.e. no “merge master into…” commits (use git rebase).

**Comment by kousu** [comment]
Yes, of course. Give me a few minutes.

Sorry, github notified me on "Of course. Enjoy your vacation!" but not on your request, so I didn't see it until just now.

**Pull request merged**